### PR TITLE
bpo-40630: add tracemalloc.reset_peak

### PR DESCRIPTION
**Pull request by huonw**
The reset_peak function sets the peak memory size to the current size,
representing a resetting of that metric. This allows for recording the
peak of specific sections of code, ignoring other code that may have
had a higher peak (since the most recent `tracemalloc.start()` or
tracemalloc.clear_traces()` call).

https://bugs.python.org/issue40630

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@huonw

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

File: Doc/library/tracemalloc.rst
Comment (vstinner) [inline Doc/library/tracemalloc.rst]: I'm not sure that this explanation belongs to the "reference API documentation". Maybe add an example showing how to get the new peak in the following section?
https://docs.python.org/dev/library/tracemalloc.html#examples

File: Lib/test/test_tracemalloc.py
Comment (vstinner) [inline Lib/test/test_tracemalloc.py:249]: I expected a test similar to https://mail.python.org/archives/list/python-ideas@python.org/thread/QDWI37A4TJXOYUKULGPY2GKD7IG2JNDC/ use case:

* Allocate large memory block (ex: 10 MB)
* Allocate small memory block (ex: 1 MB)
* Check that peak is greater than or equal to the large block
* Free the large block
* Reset peak
* Check that peak is smaller than the large block

I would prefer assertIn() which is looks fragile depending on the exact amount of memory allocated internally by Python.

Comment (huonw) [reply Lib/test/test_tracemalloc.py:249]: Hm, I'm not sure I understand. I think the test is already doing basically that, except being slightly more precise about the sizes:

1. line 263 checks the peak size is what's expected
2. line 270 checks the peak size is smaller (~equal to the current size)

Just so I understand, are you saying you would be happier with a smaller/simpler test?

> I would prefer assertIn() which is looks fragile depending on the exact amount of memory allocated internally by Python.

Sorry, I'm a bit confused by this too: did you mean "I would prefer not using `assertIn()`?

I used `range` + `assertIn` as "all in one" version of the pairs of `assertGreaterEqual(x, expected)` & `assertLessEqual(x - expected, max_error)` calls in the `test_get_traced_memory` test, and I think it's equivalent (except for being excluding the `expected + max_error` endpoint). Are you concerned that the `max_error` approach is too fragile? I could increase `max_error` to 10k or more?

File: Doc/library/tracemalloc.rst
Comment (huonw) [reply Doc/library/tracemalloc.rst]: Done!

**Comment by huonw** [comment]
Thanks for the review @vstinner 😄

File: Lib/test/test_tracemalloc.py
Comment (vstinner) [reply Lib/test/test_tracemalloc.py:249]: Yeah, max_error looks fragile. The test doesn't have to check for exact number of bytes, only that new peak is greater or less than the previous peak.

Here is a test based on your example and test_get_traced_memory():

```
    def test_reset_peak(self):
        # Python allocates some internals objects, so the test must tolerate
        # a small difference between the expected size and the real usage
        tracemalloc.clear_traces()

        # Example: allocate a large piece of memory, temporarily
        large_sum = sum(list(range(100000)))
        size1, peak1 = tracemalloc.get_traced_memory()

        # reset_peak() resets peak to traced memory: peak2 < peak1
        tracemalloc.reset_peak()
        size2, peak2 = tracemalloc.get_traced_memory()
        self.assertGreaterEqual(peak2, size2)
        self.assertLess(peak2, peak1)

        # check that peak continue to be updated if new memory is allocated:
        # peak3 > peak2
        obj_size = 1024 * 1024
        obj, obj_traceback = allocate_bytes(obj_size)
        size3, peak3 = tracemalloc.get_traced_memory()
        self.assertGreaterEqual(peak3, size3)
        self.assertGreater(peak3, peak2)
        self.assertGreaterEqual(peak3 - peak2, obj_size)
```

I use `assertGreaterEqual(peak, size)` since it's ok if the peak is greater than the traced size. Python can allocates memory in our back, there are tons of internal caches.

`self.assertGreaterEqual(peak3 - peak2, obj_size)` looks safe: we ensure that obj_size contributes to the new peak value.

Comment (huonw) [reply Lib/test/test_tracemalloc.py:249]: Ok, I've gone with your suggested test.

Would it make sense to update `test_get_traced_memory` to avoid using the `max_error` technique, at some point too? (Not in this PR.)

File: Doc/library/tracemalloc.rst
Comment (huonw) [reply Doc/library/tracemalloc.rst]: Ah, after doing this, I realised that it might be better to be a more general example. I've changed it to "Record the current and peak size of all traced memory blocks", that includes more discussion of `get_traced_memory` itself.

File: Modules/_tracemalloc.c
Comment (vstinner) [inline Modules/_tracemalloc.c:1650]: I suggest to add (here and in the doc):

```
Do nothing if the tracemalloc module is not tracing memory allocations.
```

In tracemalloc.rst, you can use:

```
:mod:`tracemalloc`
```

Comment (huonw) [reply Modules/_tracemalloc.c:1650]: I've done so. 👍

**Review by vstinner — COMMENTED** [review]
Can you please add a new "tracemalloc" sub-section in Doc/whatsnew/3.10.rst, in the "Improved Modules" section, and describe your change? You can copy/paste the NEWS entry and add (Contributed by ...): see Doc/whatsnew/3.9.rst for examples.

**Review by vstinner — APPROVED** [review]
LGTM but I would prefer to update directly What's New in Python 3.10.

**Comment by huonw** [comment]
I've added it. It's unfortunate that the review cycles meant this missed the 3.9 beta cut off, and so will presumably have to wait a year+ to be released, despite being such a small improvement.

File: Doc/library/tracemalloc.rst
Comment (vstinner) [inline Doc/library/tracemalloc.rst:347]: Oh, you need to add ".. versionadded:: 3.10".

**Review by vstinner — APPROVED** [review]
LGTM, apart the missing "versionadded" markup.

@ambv: @huonw would like to get it in Python 3.9, but we missed the feature freeze by 2 days. IMO such feature can wait for Python 3.10. On the other side, there is (very) low risk of regression. It's up to you, Łukasz. I'm waiting for @ambv feedback before merging, since currently it's documented at new feature in Python 3.10.

File: Modules/_tracemalloc.c
Comment (vstinner) [reply Modules/_tracemalloc.c:1650]: Thanks.

File: Doc/library/tracemalloc.rst
Comment (vstinner) [reply Doc/library/tracemalloc.rst]: It's now better with an example! Thanks.

File: Doc/library/tracemalloc.rst
Comment (huonw) [reply Doc/library/tracemalloc.rst:347]: Done.

**Comment by huonw** [comment]
It's fine to land for 3.10; I was just a little disconcerted and unhappy that we missed the feature cut-off by days, and so will probably have to wait for an extra year for this to be released.

**Comment by vstinner** [comment]
Thanks, I merged your PR into master.

The good news if that Python 3.10 will be the first release with a short dev cycle of only 1 year except of 18 months.

**Comment by ambv** [comment]
I would be fine for this to land in 3.9b2.

**Comment by vstinner** [comment]
@ambv: "I would be fine for this to land in 3.9b2."

Oh ok. @huonw: Do you want to backport this change to 3.9? Copy/paste-like but replaced 3.10 with 3.9. And then we can cleanup the master branch to replace 3.10 with 3.9 as well.

**Comment by huonw** [comment]
Thanks @ambv and @vstinner, I've opened #20545 and #20546 with updates for 3.9.

**Pull request merged**
